# Python基础关键字和语法

## Hello Wrold!

In [1]:
print('Hello World!')
print("Hello World!")

Hello World!
Hello World!


In [5]:
print('The \n makes a new line')
print('换\n行')

The 
 makes a new line
换
行


In [3]:
print('The \t is a tab')

The 	 is a tab


In [9]:
print('I\'m going to the movies')
print("这里有单引号''")
print('这里有双引号""')
print('这里有单引号\'')
print('这里有斜杠\\')

I'm going to the movies
这里有单引号''
这里有双引号""
这里有单引号'
这里有斜杠\


print('This is a string enclosed by \"\" not \'\'')

## 字符串操作
> Python万物皆对象
* 什么是对象
面向对象(Object Oriented)是软件开发方法，一种编程范式。面向对象的概念和应用已超越了程序设计和软件开发，扩展到如数据库系统、交互式界面、应用结构、应用平台、分布式系统、网络管理结构、CAD技术、人工智能等领域。面向对象是一种对现实世界理解和抽象的方法，是计算机编程技术发展到一定阶段后的产物。
面向对象是相对于面向过程来讲的，面向对象方法，把相关的数据和方法组织为一个整体来看待，从更高的层次来进行系统建模，更贴近事物的自然运行模式。

对象的含义是指具体的某一个事物，即在现实生活中能够看得见摸得着的事物。在面向对象程序设计中，对象所指的是计算机系统中的某一个成分。在面向对象程序设计中，对象包含两个含义，其中一个是数据，另外一个是动作。对象则是数据和动作的结合体。对象不仅能够进行操作，同时还能够及时记录下操作结果。 
方法是指对象能够进行的操作，方法同时还有另外一个名称，叫做函数。方法是类中的定义函数，其具体的作用就是对对象进行描述操作。 [3] 
继承简单地说就是一种层次模型，这种层次模型能够被重用。层次结构的上层具有通用性，但是下层结构则具有特殊性。在继承的过程中类则可以从最顶层的部分继承一些方法和变量。类除了可以继承以外同时还能够进行修改或者添加。通过这样的方式能够有效提高工作效率。在这里举一个例子，当类X继承了类Y后，此时的类X则是一个派生类，而类Y属于一个基类。 [3]  继承是从一般演绎到特殊的过程，可以减少知识表示的冗余内容，知识库的维护和修正都非常方便。更有利于衍生复杂的系统。 [4] 
类是具有相同特性（数据元素）和行为（功能）的对象的抽象。因此，对象的抽象是类，类的具体化就是对象，也可以说类的实例是对象，类实际上就是一种数据类型。类具有属性，它是对象的状态的抽象，用数据结构来描述类的属性。类具有操作，它是对象的行为的抽象，用操作名和实现该操作的方法来描述。 [2]  类映射的每一个对象都具有这些数据和操作方法，类的继承具有层次性和结构性，高层次对象封装复杂行为，具体细节对该层次知识保持透明，可以减小问题求解的复杂度。
封装是将数据和代码捆绑到一起，对象的某些数据和代码可以是私有的，不能被外界访问，以此实现对数据和代码不同级别的访问权限。防止了程序相互依赖性而带来的变动影响，面向对象的封装比传统语言的封装更为清晰、更为有力。有效实现了两个目标：对数据和行为的包装和信息隐藏。 [1] 
多态是指不同事物具有不同表现形式的能力。多态机制使具有不同内部结构的对象可以共享相同的外部接口，通过这种方式减少代码的复杂度。一个接口，多种方式。 [1] 
动态绑定指的是将一个过程调用与相应代码链接起来的行为。动态绑定是指与给定的过程调用相关联的代码只有在运行期才可知的一种绑定，它是多态实现的具体形式。 [1] 
消息传递：对象之间需要相互沟通，沟通的途径就是对象之间收发信息。消息内容包括接收消息的对象的标识，需要调用的函数的标识，以及必要的信息。消息传递的概念使得对现实世界的描述更容易。 [1] 
面向对象的方法就是利用抽象、封装等机制，借助于对象、类、继承、消息传递等概念进行软件系统构造的软件开发方法。

In [10]:
firstVariable = 'Hello World!'

In [15]:
a='123 456'

In [14]:
firstVariable.split(' ')

['Hello', 'World!']

In [7]:
print(firstVariable.lower())
print(firstVariable.upper())
print(firstVariable.title())

hello world!
HELLO WORLD!
Hello World!


In [8]:
firstVariable.split(' ')

['Hello', 'World!']

In [16]:
b=firstVariable.split(' ')

In [20]:
b

['Hello', 'World!']

In [9]:
' '.join(['Hello','World!'])

'Hello World!'

In [21]:
'，'.join(b)

'Hello，World!'

In [22]:
print('0'+'1')
print('Hello'+"World!")

01
HelloWorld!


In [23]:
'0'*3
firstVariable*4

'Hello World!Hello World!Hello World!Hello World!'

In [2]:
'Fizz'+'Buzz'

'FizzBuzz'

In [26]:
a='l'
b='l'
'hel'+a+'o'+' '+'wor'+b+'d'

'hello world'

In [3]:
'{}{}'.format('Fizz', 'Buzz')

'FizzBuzz'

In [25]:
'hel{}o wor{}d'.format('5','l')

'hel5o world'

## 基础数学

In [4]:
1+1

2

In [5]:
130-2.0

128.0

In [6]:
130-2

128

In [7]:
130/2

65.0

In [8]:
130.0/2

65.0

In [9]:
2*3

6

In [10]:
2**3

8

In [11]:
9%3

0

In [12]:
8%3

2

## if语句

Python的语句块使用四个空格或缩进符代替

等于==，大于>，不等于!=，&与，|或

In [27]:
num = 3
if num == 3:
    print(num)

3


In [14]:
num = 3
if num > 10:
    print(num)

In [15]:
num = 3
if num % 3 == 0:
    print("Fizz")

Fizz


In [16]:
num = 10
if num % 5 == 0:
    print("Buzz")

Buzz


In [17]:
if True:
    print("This was True")

This was True


In [18]:
if False: 
    print("Nothing printed")

In [19]:
num = 4
num > 0 and num  < 15

True

In [32]:
True | False

True

In [20]:
if num > 0 and num  < 15:
    print(num)

4


In [21]:
num = 4
num > 0 or num  > 15

True

In [22]:
if num > 0 or num  > 15:
    print(num)

4


In [23]:
if False or False:
    print('Nothing will print out')

In [24]:
num = 10
not num < 20 

False

## else语句

In [25]:
num = 1
if num > 3 :
    print("Hi")

In [26]:
num = 1
if num > 3 :
    print("Hi")
else: 
    print("number is not greater than 3")

number is not greater than 3


In [27]:
num = 4
if num > 3 :
    print("Hi")
else: 
    print("number is not greater than 3")

Hi


In [28]:
num = 3
if num % 2 == 0:
    print("Your integer is even")
else: 
    print("Your integer is odd")

Your integer is odd


## elif语句

In [29]:
num = 21
if num > 50:
    print('num is larger than 50')
elif num == 21:
    print('num = 21')
else:
    print('Catchall condition')

num = 21


In [36]:
num=60
if num>=90:
    print('优')
elif num>=60 and num<90:
    print('良')
else:
    print('差')

良


In [30]:
my_num = 5
if my_num % 2 == 0:
    print("Your number is even")
elif my_num % 2 != 0:
    print("Your number is odd")
else: 
    print("Are you sure your number is an integer?")

Your number is odd


In [38]:
dice_value = 3
if dice_value == 1:
    print('You rolled a {}. Great job!'.format(dice_value))
elif dice_value == 2:
    print('You rolled a {}. Great job!'.format(dice_value))
elif dice_value == 3:
    print('You rolled a {}. Great job!'.format(dice_value))
elif dice_value == 4:
    print('You rolled a {}. Great job!'.format(dice_value))
elif dice_value == 5:
    print('You rolled a {}. Great job!'.format(dice_value))
elif dice_value == 6:
    print('You rolled a {}. Great job!'.format(dice_value))
else:
    print('None of the conditions above (if elif) were evaluated as True')

You rolled a 3. Great job!


In [32]:
num = 10

In [33]:
if num % 3 == 0 and num % 5 == 0:
    print('FizzBuzz')
elif num % 3 == 0:
    print('Fizz')
elif num % 5 == 0:
    print('Buzz')
else:
    print(str(num))

Buzz


In [34]:
num = 15

In [35]:
string = ""
if num % 3 == 0:
    string = string + "Fizz"
if num % 5 == 0:
    string = string + "Buzz"
if num % 5 != 0 and num % 3 != 0:
    string = string + str(num)
print(string)

FizzBuzz


## 列表
List

### 访问列表里面的值

In [44]:
z = [3, 7, 4, 2, 8]

In [45]:
len(z)

5

使用下标取值

In [37]:
z[0]

3

In [46]:
z[4]

8

In [42]:
z[-2]

4

In [41]:
z[-1]

2

### 列表切片

[从第几个到第几个]半集合[)

In [40]:
z[0:2]

[3, 7]

In [41]:
z[:3]

[3, 7, 4]

In [42]:
z[1:]

[7, 4, 2]

### 取列表的最大值, 最小值, 长度, 以及总和

In [43]:
print(min(z), max(z), len(z), sum(z))

2 7 4 16


min(),max()是Python的内置函数

### 对列表中对象出现次数进行统计

In [47]:
random_list = [4, 1, 5, 4, 10, 4]
random_list.count(4)

3

In [51]:
help(random_list.count)

Help on built-in function count:

count(value, /) method of builtins.list instance
    Return number of occurrences of value.



### 返回列表第一个指针

In [52]:
random_list.index(4)

0

In [55]:
help(random_list.index)

Help on built-in function index:

index(value, start=0, stop=9223372036854775807, /) method of builtins.list instance
    Return first index of value.
    
    Raises ValueError if the value is not present.



In [56]:
random_list.index(4, 3)

3

In [47]:
random_list.index(4, 5, 6)

5

### 对列表进行排序

In [82]:
x = [3, 7, 2, 11, 8, 10, 4]
y = ['Steve', 'Rachel', 'Michael', 'Adam', 'Monica', 'Jessica', 'Lester']

In [66]:
z=sorted(x)

In [67]:
z

[2, 3, 4, 7, 8, 10, 11]

In [68]:
x

[3, 7, 2, 11, 8, 10, 4]

In [58]:
x.sort()
print(x)

[2, 3, 4, 7, 8, 10, 11]


In [59]:
help(x.sort)

Help on built-in function sort:

sort(*, key=None, reverse=False) method of builtins.list instance
    Sort the list in ascending order and return None.
    
    The sort is in-place (i.e. the list itself is modified) and stable (i.e. the
    order of two equal elements is maintained).
    
    If a key function is given, apply it once to each list item and sort them,
    ascending or descending, according to their function values.
    
    The reverse flag can be set to sort in descending order.



In [50]:
x.sort(reverse = True)
print(x)

[11, 10, 8, 7, 4, 3, 2]


In [51]:
y.sort()
print(y)

['Adam', 'Jessica', 'Lester', 'Michael', 'Monica', 'Rachel', 'Steve']


In [52]:
y.sort(reverse = True)
print(y)

['Steve', 'Rachel', 'Monica', 'Michael', 'Lester', 'Jessica', 'Adam']


In [53]:
new_list = sorted(y)
new_list

['Adam', 'Jessica', 'Lester', 'Michael', 'Monica', 'Rachel', 'Steve']

In [54]:
y

['Steve', 'Rachel', 'Monica', 'Michael', 'Lester', 'Jessica', 'Adam']

### 在列表结尾添加一个对象

In [70]:
x

[3, 7, 2, 11, 8, 10, 4]

In [69]:
help(x)

Help on list object:

class list(object)
 |  list(iterable=(), /)
 |  
 |  Built-in mutable sequence.
 |  
 |  If no argument is given, the constructor creates a new empty list.
 |  The argument must be an iterable if specified.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(...)
 |      x.__getitem__(y) <==> x[y]
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __iadd__(self, value, /)
 |      Implement self+=value.
 |  
 |  __imul__(self, value, /)
 |      Implement self*=value.
 |  
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self)) for accurate sign

In [73]:
x.append(3)
print(x)

[3, 7, 2, 11, 8, 10, 4, 3, 3, 3]


### 删除列表中一个对象

In [57]:
x.remove(10)
print(x)

[11, 8, 7, 4, 3, 2, 3]


### 删除列表中指定位置的对象

In [58]:
print(x)

[11, 8, 7, 4, 3, 2, 3]


In [59]:
x.pop(3)#括号内是index

4

In [60]:
print(x)

[11, 8, 7, 3, 2, 3]


### 合并列表

In [75]:
x.extend(y)

In [61]:
x.extend([4, 5])

In [83]:
print(x+y)

[3, 7, 2, 11, 8, 10, 4, 'Steve', 'Rachel', 'Michael', 'Adam', 'Monica', 'Jessica', 'Lester']


In [84]:
x

[3, 7, 2, 11, 8, 10, 4]

In [78]:
print('x+y=', x+y)

x+y= [3, 7, 2, 11, 8, 10, 4, 'Steve', 'Rachel', 'Michael', 'Adam', 'Monica', 'Jessica', 'Lester', 'Steve', 'Rachel', 'Michael', 'Adam', 'Monica', 'Jessica', 'Lester']


### 在列表指定位置前插入对象

In [87]:
x

[3, 7, 2, 11, 8, 10, 4, 'hello']

In [90]:
x[0]+x[7]

'3hello'

In [86]:
x.append('hello')

In [91]:
x.insert(4, [4, 5])

In [92]:
x

[3, 7, 2, 11, [4, 5], 8, 10, 4, 'hello']

In [97]:
x[4].insert(1,['a','b'])

In [98]:
x

[3, 7, 2, 11, [4, ['a', 'b'], 5], 8, 10, 4, 'hello']

# Python中常见的数据结构

## 数字

In [2]:
a = 1
b = 3.14
c = 1.23j
print(type(a), type(b), type(c))

<class 'int'> <class 'float'> <class 'complex'>


In [4]:
int(3.8)

3

In [5]:
float(3)

3.0

In [6]:
type(1+3.1)

float

In [2]:
type(5*0.6)

float

## 布尔

In [3]:
a = True
b = False

In [4]:
type(a)

bool

In [5]:
a+b

1

In [6]:
True+True

2

## 字典

### 定义字典

字典定义{'键key':'值value'}，键值对

In [140]:
webstersDict = {'person': 'a human being, whether an adult or child', 'marathon': 'a running race that is about 26 miles',
                'resist': ' to remain strong against the force or effect of (something)', 'run': 'to move with haste; act quickly'}

In [105]:
webstersDict

{'person': 'a human being, whether an adult or child',
 'marathon': 'a running race that is about 26 miles',
 'resist': ' to remain strong against the force or effect of (something)',
 'run': 'to move with haste; act quickly'}

### 访问字典中的值

In [101]:
webstersDict['marathon']

'a running race that is about 26 miles'

Help on dict object:

class dict(object)
 |  dict() -> new empty dictionary
 |  dict(mapping) -> new dictionary initialized from a mapping object's
 |      (key, value) pairs
 |  dict(iterable) -> new dictionary initialized as if via:
 |      d = {}
 |      for k, v in iterable:
 |          d[k] = v
 |  dict(**kwargs) -> new dictionary initialized with the name=value pairs
 |      in the keyword argument list.  For example:  dict(one=1, two=2)
 |  
 |  Built-in subclasses:
 |      StgDict
 |  
 |  Methods defined here:
 |  
 |  __contains__(self, key, /)
 |      True if the dictionary has the specified key, else False.
 |  
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(...)
 |      x.__getitem__(y) <==> x[y]
 |  
 |  __gt__(self, value, /)
 |      Return self>va

### 更新字典

In [109]:
webstersDict['shoe'] = 'an external covering for the human foot'
webstersDict['shoe']

'an external covering for the human foot'

In [115]:
webstersDict['shoe']

'穿在脚上的东西'

In [111]:
webstersDict['shoe'] = '穿在脚上的东西'

In [113]:
webstersDict.update({'shirt': '上身穿的衣物'})

In [108]:
webstersDict.update({'shirt': 'a long- or short-sleeved garment for the upper part of the body',
                     'shoe': 'an external covering for the human foot, usually of leather and consisting of a more or less stiff or heavy sole and a lighter upper part ending a short distance above, at, or below the ankle.'})
webstersDict

{'person': 'a human being, whether an adult or child',
 'marathon': 'a running race that is about 26 miles',
 'resist': ' to remain strong against the force or effect of (something)',
 'run': 'to move with haste; act quickly',
 'shoe': 'an external covering for the human foot, usually of leather and consisting of a more or less stiff or heavy sole and a lighter upper part ending a short distance above, at, or below the ankle.',
 'shirt': 'a long- or short-sleeved garment for the upper part of the body'}

In [116]:
del webstersDict['resist']
webstersDict

{'person': 'a human being, whether an adult or child',
 'marathon': 'a running race that is about 26 miles',
 'run': 'to move with haste; act quickly',
 'shoe': '穿在脚上的东西',
 'shirt': '上身穿的衣物'}

In [117]:
webstersDict[1]='1'

In [118]:
webstersDict

{'person': 'a human being, whether an adult or child',
 'marathon': 'a running race that is about 26 miles',
 'run': 'to move with haste; act quickly',
 'shoe': '穿在脚上的东西',
 'shirt': '上身穿的衣物',
 1: '1'}

### 不是所有东西都可以当作Key

In [74]:
webstersDict[['sock']] = 'a short stocking usually reaching to the calf or just above the ankle.'

TypeError: unhashable type: 'list'

### 使用get()方法返回给定键的值

In [119]:
storyCount = {'is': 100, 'the': 90, 'Michael': 12, 'runs': 5}
storyCount

{'is': 100, 'the': 90, 'Michael': 12, 'runs': 5}

In [120]:
storyCount['runs']

5

In [123]:
storyCount.get('Michael', 0)

12

In [121]:
help(storyCount.get)

Help on built-in function get:

get(key, default=None, /) method of builtins.dict instance
    Return the value for key if key is in the dictionary, else default.



In [124]:
print(storyCount.get('run'))

None


In [126]:
print(storyCount.get('run', '找不到这个键'))

找不到这个键


### 删除键，但同时可以返回值

In [127]:
count = storyCount.pop('the')
print(count)

90


In [128]:
storyCount

{'is': 100, 'Michael': 12, 'runs': 5}

### 遍历字典

In [137]:
print(storyCount.keys())
print(storyCount.values())

dict_keys(['is', 'Michael', 'runs'])
dict_values([100, 12, 5])


In [150]:
x=[1,2,3,4,5]
for i in x:
    print(i)

1
2
3
4
5


In [151]:
webstersDict

{'person': 'a human being, whether an adult or child',
 'marathon': 'a running race that is about 26 miles',
 'resist': ' to remain strong against the force or effect of (something)',
 'run': 'to move with haste; act quickly'}

In [156]:
webstersDict['person']

'a human being, whether an adult or child'

In [159]:
for key in webstersDict:
    print(key+'：'+webstersDict[key])

person：a human being, whether an adult or child
marathon：a running race that is about 26 miles
resist： to remain strong against the force or effect of (something)
run：to move with haste; act quickly


In [161]:
for key in webstersDict:
    print('{}任意字符{}'.format(key, webstersDict[key]))
#     print(key)
#     print(webstersDict[key])

person任意字符a human being, whether an adult or child
marathon任意字符a running race that is about 26 miles
resist任意字符 to remain strong against the force or effect of (something)
run任意字符to move with haste; act quickly


In [144]:
webstersDict.items()

dict_items([('person', 'a human being, whether an adult or child'), ('marathon', 'a running race that is about 26 miles'), ('resist', ' to remain strong against the force or effect of (something)'), ('run', 'to move with haste; act quickly')])

In [146]:
for key, value in webstersDict.items():
    print('{}：{}'.format(key,value))
#     print(key, value)

person：a human being, whether an adult or child
marathon：a running race that is about 26 miles
resist： to remain strong against the force or effect of (something)
run：to move with haste; act quickly


## 元组

### 初始化一个元组

In [1]:
emptyTuple = ()

In [2]:
emptyTuple = tuple()

In [3]:
z = (3, 7, 4, 2)
z = 3, 7, 4, 2

In [ ]:
x = [3, 4, 5, 3]

In [4]:
tup1 = ('Michael',)
tup2 = 'Michael', 
notTuple = ('Michael')

### 访问元组内的值

In [5]:
z = (3, 7, 4, 2)
print(z[0])

3


In [7]:
z[-1]

2

In [97]:
print(z[-1])

2


In [98]:
print(z[3])

2


### 元组切片

In [8]:
z = (3, 7, 4, 2)
print(z[0:2])

(3, 7)


In [9]:
z[:2]

(3, 7)

In [10]:
z[2:]

(4, 2)

In [100]:
print(z[:3])

(3, 7, 4)


In [101]:
print(z[-4:-1])

(3, 7, 4)


### 元组是不可改变的

In [102]:
z = (3, 7, 4, 2)
z[1] = "fish"

TypeError: 'tuple' object does not support item assignment

In [12]:
x = [3, 7, 4, 2]
x[1]=6
print(x)

[3, 6, 4, 2]


In [16]:
tup1 = ('Python', 'SQL')
tup2 = ('R',)
new_tuple = tup1 + tup2;
print(new_tuple)

('Python', 'SQL', 'R')


In [18]:
z+tup1

(3, 7, 4, 2, 'Python', 'SQL')

In [21]:
dir(z)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'count',
 'index']

### Tuple方法

In [22]:
animals = ('lama', 'sheep', 'lama', 48)

#### index 方法(索引)

In [25]:
print(animals.index('lama'))

0


#### count 方法(计数)

In [26]:
print(animals.count('lama'))

2


### 遍历元组

In [107]:
for item in ('lama', 'sheep', 'lama', 48):
    print(item)

lama
sheep
lama
48


In [27]:
for item in ['lama', 'sheep', 'lama', 48]:
    print(item)

lama
sheep
lama
48


### 元组拆包

In [28]:
x, y = (7, 10)
print("Value of x is {}, the value of y is {}.".format(x, y))

Value of x is 7, the value of y is 10.


In [29]:
a,b,c,d=('lama', 'sheep', 'lama', 48)

In [32]:
a,b,c,d=['lama', 'sheep', 'lama', 48]

### 枚举

In [109]:
friends = ('Steve', 'Rachel', 'Michael', 'Monica')
for index, friend in enumerate(friends):
    print(index, friend)

0 Steve
1 Rachel
2 Michael
3 Monica


### 元组相对列表的优势

#### 迭代速度快

In [35]:
import timeit
print('Tuple time: ', timeit.timeit('x=(1,2,3,4,5,6,7,8,9,10,11,12)', number=1000000))
print('List time: ', timeit.timeit('x=[1,2,3,4,5,6,7,8,9,10,11,12]', number=1000000))

Tuple time:  0.06974570000011227
List time:  0.29053109999995286


#### 元组可以用作字典键

In [36]:
bigramsTupleDict = {('this', 'is'): 23,
                    ('is', 'a'): 12,
                    ('a', 'sentence'): 2}
print(bigramsTupleDict)

{('this', 'is'): 23, ('is', 'a'): 12, ('a', 'sentence'): 2}


In [37]:
bigramsTupleDict[('this', 'is')]

23

#### 列表不可以用作字典键

In [38]:
bigramsListDict = {['this', 'is']: 23,
                   ['is', 'a']: 12,
                   ['a', 'sentence']: 2}
print(bigramsListDict)

TypeError: unhashable type: 'list'

#### 元组可以是集合中的值

In [39]:
graphicDesigner = {('this', 'is'),
                   ('is', 'a'),
                   ('a', 'sentence')}
print(graphicDesigner)

{('is', 'a'), ('this', 'is'), ('a', 'sentence')}


#### 列表不可以是集合中的值

In [40]:
graphicDesigner = {['this', 'is'],
                   ['is', 'a'],
                   ['a', 'sentence']}
print(graphicDesigner)

TypeError: unhashable type: 'list'

## Task: 用Python生成斐波那契序列

Fibonacci序列是一个整数序列，其特征在于前两个之后的每个数字是前两个数字的总和。根据定义，Fibonacci序列中的前两个数字是1和1，或0和1，具体取决于所选择的序列起点，以及每个后续数字是前两个数字的总和。

In [115]:
print(1, 1, 2, 3, 5, 8, 13, 21, 34, 55)

1 1 2 3 5 8 13 21 34 55


In [46]:
x=tuple(range(10))

In [47]:
x

(0, 1, 2, 3, 4, 5, 6, 7, 8, 9)

In [49]:
for i in range(11):
    print(i)

0
1
2
3
4
5
6
7
8
9
10


In [116]:
a, b = 1, 1
for i in range(10):
    print("Fib(a): ", a, "b is: ", b)
    a=b
    b=a+b

Fib(a):  1 b is:  1
Fib(a):  1 b is:  2
Fib(a):  2 b is:  3
Fib(a):  3 b is:  5
Fib(a):  5 b is:  8
Fib(a):  8 b is:  13
Fib(a):  13 b is:  21
Fib(a):  21 b is:  34
Fib(a):  34 b is:  55
Fib(a):  55 b is:  89


# for循环

In [50]:
for number in [23, 41, 12, 16, 7]:
    print(number)
print('Hi')

23
41
12
16
7
Hi


## 枚举

In [51]:
friends = ['steve', 'rachel', 'michael', 'adam', 'monica']
for index, friend in enumerate(friends):
    print(index, friend)

0 steve
1 rachel
2 michael
3 adam
4 monica


In [55]:
for i in enumerate(friends):
    a,b=i
    print(a,b)

0 steve
1 rachel
2 michael
3 adam
4 monica


In [56]:
for a,b in enumerate(friends):
    print(a,b)

0 steve
1 rachel
2 michael
3 adam
4 monica


## Task

In [74]:
text = '''On a dark desert highway, cool wind in my hair Warm smell of colitas, rising up through the air Up ahead in the distance, I saw a shimmering light My head grew heavy and my sight grew dim I had to stop for the night There she stood in the doorway; I heard the mission bell And I was thinking to myself, "This could be Heaven or this could be Hell" Then she lit up a candle and she showed me the way'''
print(text)
for char in '-.,;\n"\'':
    text = text.replace(char,' ')
print(text)

On a dark desert highway, cool wind in my hair Warm smell of colitas, rising up through the air Up ahead in the distance, I saw a shimmering light My head grew heavy and my sight grew dim I had to stop for the night There she stood in the doorway; I heard the mission bell And I was thinking to myself, "This could be Heaven or this could be Hell" Then she lit up a candle and she showed me the way
On a dark desert highway  cool wind in my hair Warm smell of colitas  rising up through the air Up ahead in the distance  I saw a shimmering light My head grew heavy and my sight grew dim I had to stop for the night There she stood in the doorway  I heard the mission bell And I was thinking to myself   This could be Heaven or this could be Hell  Then she lit up a candle and she showed me the way


In [66]:
for char in '-.,;\n"\'':
    print(char)

-
.
,
;


"
'


In [71]:
new=text.replace('a','b')

In [73]:
text

'On a dark desert highway, cool wind in my hair Warm smell of colitas, rising up through the air Up ahead in the distance, I saw a shimmering light My head grew heavy and my sight grew dim I had to stop for the night There she stood in the doorway; I heard the mission bell And I was thinking to myself, "This could be Heaven or this could be Hell" Then she lit up a candle and she showed me the way'

In [67]:
help(text.replace)

Help on built-in function replace:

replace(old, new, count=-1, /) method of builtins.str instance
    Return a copy with all occurrences of substring old replaced by new.
    
      count
        Maximum number of occurrences to replace.
        -1 (the default value) means replace all occurrences.
    
    If the optional argument count is given, only the first count occurrences are
    replaced.



In [59]:
text = 'On a dark desert highway, cool wind in my hair Warm smell of colitas, rising up through the air Up ahead in the distance, I saw a shimmering light My head grew heavy and my sight grew dim I had to stop for the night There she stood in the doorway; I heard the mission bell And I was thinking to myself, "This could be Heaven or this could be Hell" Then she lit up a candle and she showed me the way'

In [65]:
text[26:30]

'cool'

In [64]:
for i,j in enumerate(text):
    print(i,j)

0 O
1 n
2  
3 a
4  
5 d
6 a
7 r
8 k
9  
10 d
11 e
12 s
13 e
14 r
15 t
16  
17 h
18 i
19 g
20 h
21 w
22 a
23 y
24 ,
25  
26 c
27 o
28 o
29 l
30  
31 w
32 i
33 n
34 d
35  
36 i
37 n
38  
39 m
40 y
41  
42 h
43 a
44 i
45 r
46  
47 W
48 a
49 r
50 m
51  
52 s
53 m
54 e
55 l
56 l
57  
58 o
59 f
60  
61 c
62 o
63 l
64 i
65 t
66 a
67 s
68 ,
69  
70 r
71 i
72 s
73 i
74 n
75 g
76  
77 u
78 p
79  
80 t
81 h
82 r
83 o
84 u
85 g
86 h
87  
88 t
89 h
90 e
91  
92 a
93 i
94 r
95  
96 U
97 p
98  
99 a
100 h
101 e
102 a
103 d
104  
105 i
106 n
107  
108 t
109 h
110 e
111  
112 d
113 i
114 s
115 t
116 a
117 n
118 c
119 e
120 ,
121  
122 I
123  
124 s
125 a
126 w
127  
128 a
129  
130 s
131 h
132 i
133 m
134 m
135 e
136 r
137 i
138 n
139 g
140  
141 l
142 i
143 g
144 h
145 t
146  
147 M
148 y
149  
150 h
151 e
152 a
153 d
154  
155 g
156 r
157 e
158 w
159  
160 h
161 e
162 a
163 v
164 y
165  
166 a
167 n
168 d
169  
170 m
171 y
172  
173 s
174 i
175 g
176 h
177 t
178  
179 g
180 r
181 e
182 w
183  
184 d


In [75]:
text

'On a dark desert highway  cool wind in my hair Warm smell of colitas  rising up through the air Up ahead in the distance  I saw a shimmering light My head grew heavy and my sight grew dim I had to stop for the night There she stood in the doorway  I heard the mission bell And I was thinking to myself   This could be Heaven or this could be Hell  Then she lit up a candle and she showed me the way'

In [76]:
help(text.split)

Help on built-in function split:

split(sep=None, maxsplit=-1) method of builtins.str instance
    Return a list of the words in the string, using sep as the delimiter string.
    
    sep
      The delimiter according which to split the string.
      None (the default value) means split according to any whitespace,
      and discard empty strings from the result.
    maxsplit
      Maximum number of splits to do.
      -1 (the default value) means no limit.



In [78]:
print(text.split(' '))

['On', 'a', 'dark', 'desert', 'highway', '', 'cool', 'wind', 'in', 'my', 'hair', 'Warm', 'smell', 'of', 'colitas', '', 'rising', 'up', 'through', 'the', 'air', 'Up', 'ahead', 'in', 'the', 'distance', '', 'I', 'saw', 'a', 'shimmering', 'light', 'My', 'head', 'grew', 'heavy', 'and', 'my', 'sight', 'grew', 'dim', 'I', 'had', 'to', 'stop', 'for', 'the', 'night', 'There', 'she', 'stood', 'in', 'the', 'doorway', '', 'I', 'heard', 'the', 'mission', 'bell', 'And', 'I', 'was', 'thinking', 'to', 'myself', '', '', 'This', 'could', 'be', 'Heaven', 'or', 'this', 'could', 'be', 'Hell', '', 'Then', 'she', 'lit', 'up', 'a', 'candle', 'and', 'she', 'showed', 'me', 'the', 'way']


In [79]:
print(text.split(' ')[0:20])

['On', 'a', 'dark', 'desert', 'highway', '', 'cool', 'wind', 'in', 'my', 'hair', 'Warm', 'smell', 'of', 'colitas', '', 'rising', 'up', 'through', 'the']


In [80]:
print(len(text),len(text.split(' ')))

398 90


In [91]:
len('')

0

In [16]:
cleaned_list = []
for word in text.split(' '): 
    word_length = len(word)
    if word_length > 0:
        cleaned_list.append(word)

In [85]:
cleaned_list = []
for word in text.split(' '):
    if len(word)>0:
        cleaned_list.append(word)

In [88]:
print(text.split(' '))

['On', 'a', 'dark', 'desert', 'highway', '', 'cool', 'wind', 'in', 'my', 'hair', 'Warm', 'smell', 'of', 'colitas', '', 'rising', 'up', 'through', 'the', 'air', 'Up', 'ahead', 'in', 'the', 'distance', '', 'I', 'saw', 'a', 'shimmering', 'light', 'My', 'head', 'grew', 'heavy', 'and', 'my', 'sight', 'grew', 'dim', 'I', 'had', 'to', 'stop', 'for', 'the', 'night', 'There', 'she', 'stood', 'in', 'the', 'doorway', '', 'I', 'heard', 'the', 'mission', 'bell', 'And', 'I', 'was', 'thinking', 'to', 'myself', '', '', 'This', 'could', 'be', 'Heaven', 'or', 'this', 'could', 'be', 'Hell', '', 'Then', 'she', 'lit', 'up', 'a', 'candle', 'and', 'she', 'showed', 'me', 'the', 'way']


In [86]:
print(cleaned_list)

['On', 'a', 'dark', 'desert', 'highway', 'cool', 'wind', 'in', 'my', 'hair', 'Warm', 'smell', 'of', 'colitas', 'rising', 'up', 'through', 'the', 'air', 'Up', 'ahead', 'in', 'the', 'distance', 'I', 'saw', 'a', 'shimmering', 'light', 'My', 'head', 'grew', 'heavy', 'and', 'my', 'sight', 'grew', 'dim', 'I', 'had', 'to', 'stop', 'for', 'the', 'night', 'There', 'she', 'stood', 'in', 'the', 'doorway', 'I', 'heard', 'the', 'mission', 'bell', 'And', 'I', 'was', 'thinking', 'to', 'myself', 'This', 'could', 'be', 'Heaven', 'or', 'this', 'could', 'be', 'Hell', 'Then', 'she', 'lit', 'up', 'a', 'candle', 'and', 'she', 'showed', 'me', 'the', 'way']


In [89]:
len(cleaned_list)

83

In [17]:
cleaned_list[0:20]

['On',
 'a',
 'dark',
 'desert',
 'highway',
 'cool',
 'wind',
 'in',
 'my',
 'hair',
 'Warm',
 'smell',
 'of',
 'colitas',
 'rising',
 'up',
 'through',
 'the',
 'air',
 'Up']

## Continue

不执行continue下面的语句，重新开始新的循环

In [97]:
cleaned_list = []

for word in text.split(' '):
    if len(word) == 0:
        continue
    cleaned_list.append(word)

In [96]:
for i in range(10):
    if i==5:
        continue
    print(i)

0
1
2
3
4
6
7
8
9


In [98]:
print(cleaned_list)

['On', 'a', 'dark', 'desert', 'highway', 'cool', 'wind', 'in', 'my', 'hair', 'Warm', 'smell', 'of', 'colitas', 'rising', 'up', 'through', 'the', 'air', 'Up', 'ahead', 'in', 'the', 'distance', 'I', 'saw', 'a', 'shimmering', 'light', 'My', 'head', 'grew', 'heavy', 'and', 'my', 'sight', 'grew', 'dim', 'I', 'had', 'to', 'stop', 'for', 'the', 'night', 'There', 'she', 'stood', 'in', 'the', 'doorway', 'I', 'heard', 'the', 'mission', 'bell', 'And', 'I', 'was', 'thinking', 'to', 'myself', 'This', 'could', 'be', 'Heaven', 'or', 'this', 'could', 'be', 'Hell', 'Then', 'she', 'lit', 'up', 'a', 'candle', 'and', 'she', 'showed', 'me', 'the', 'way']


## Break

循环结束语句

In [99]:
cleaned_list = []
for word in text.split(' '): 
    if word == 'desert':
        print('I found the word I was looking for')
        break
    cleaned_list.append(word)
# cleaned_list

I found the word I was looking for


In [101]:
for i in range(10):
    if i==5:
        break
    print(i)

0
1
2
3
4


In [100]:
cleaned_list

['On', 'a', 'dark']

## Task (顺道介绍一下Range函数)

In [103]:
help(range)

Help on class range in module builtins:

class range(object)
 |  range(stop) -> range object
 |  range(start, stop[, step]) -> range object
 |  
 |  Return an object that produces a sequence of integers from start (inclusive)
 |  to stop (exclusive) by step.  range(i, j) produces i, i+1, i+2, ..., j-1.
 |  start defaults to 0, and stop is omitted!  range(4) produces 0, 1, 2, 3.
 |  These are exactly the valid indices for a list of 4 elements.
 |  When step is given, it specifies the increment (or decrement).
 |  
 |  Methods defined here:
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(self, key, /)
 |      Return self[key].
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __hash__(self, /)
 |

In [104]:
for i in range(10):
    print(i)

0
1
2
3
4
5
6
7
8
9


In [106]:
for i in range(5,10):
    print(i)

5
6
7
8
9


In [108]:
for i in range(2,10,2):
    print(i)

2
4
6
8


In [109]:
even_numbers = []
odd_numbers = []

for number in range(1, 51):
    if number % 2 == 0:
        even_numbers.append(number)
    else:
        odd_numbers.append(number)

%求余

In [113]:
30%20

10

In [110]:
print("Even Numbers: ", even_numbers)

Even Numbers:  [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50]


In [111]:
print("Odd Numbers: ", odd_numbers)

Odd Numbers:  [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49]


# while循环

## While 循环

In [23]:
count = 0
while count <= 5:
    print(count)
    count = count + 1

0
1
2
3
4
5


In [115]:
a = 10
while a>5:
    print(a)
    a=a-1

10
9
8
7
6


## break语句

In [117]:
count = 0
while count <= 5:
    if count == 2:
        break
    count += 1
    print (count)

1
2
3
4
5
6


In [118]:
num = 0
while True:
    if num == 2:
        print('Found 2')
        break
    num += 1
    print(num)

1
2
Found 2


In [27]:
1 % 5

1

In [28]:
5 % 5

0

In [29]:
x = 1
while x % 5 != 0:
    x += 1
    print(x)

2
3
4
5


In [119]:
candidates = list(range(0, 5))
candidates

[0, 1, 2, 3, 4]

In [123]:
tuple(range(0, 5))

(0, 1, 2, 3, 4)

In [120]:
type(range(5))

range

In [122]:
dir(range)

['__bool__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'count',
 'index',
 'start',
 'step',
 'stop']

In [124]:
while len(candidates) > 0: 
    first = candidates[0]
    candidates.remove(first)
    print(candidates)

[1, 2, 3, 4]
[2, 3, 4]
[3, 4]
[4]
[]


# 函数

## 创建函数的语法
使用函数的目的是为了代码复用

In [32]:
def function_name(parameters):
    """docstring"""
    statement(s)

In [125]:
def print_word(word=None):
    print(word)

In [128]:
print_word(1)

1


## 如何在python中调用函数？

In [132]:
name = "Emma"
def greet(name):
    print("Hello, " + name + ". Good morning!")
greet(name)

Hello, Emma. Good morning!


In [130]:
greet('曾')

Hello, 曾. Good morning!


In [134]:
help(greet)

Help on function greet in module __main__:

greet(name)



## 文档字符串

In [133]:
print(greet.__doc__)

None


## 返回语句

In [135]:
def absolute_value(num):
    if num >= 0:
        return num
    else:
        return -num
print(absolute_value(2))
print(absolute_value(-4))

2
4


In [140]:
def greet(name=''):
    return "Hello, " + name + ". Good morning!"

In [143]:
x=greet('abc')

In [144]:
x

'Hello, abc. Good morning!'

## 变量的范围和寿命

函数内部的变量是局部变量，函数外面是无法使用的，函数外部的变量是全局变量，函数内是可以使用的

In [145]:
def my_func():
    x = 10
    print("Value inside function:", x)


x = 20
my_func()
print("Value outside function:", x)

Value inside function: 10
Value outside function: 20


In [151]:
aa = 99
def my_func():
    aa=100
    print(aa)
my_func()

100


In [154]:
# 返回某个数以内的全部偶数
def even_numbers(num=0):
    even_num=[]
    for i in range(num):
        if i%2==0:
            even_num.append(i)
    return even_num

In [157]:
print(even_numbers(20))
print(even_numbers(40))
print(even_numbers(60))
print(even_numbers(80))

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]
[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38]
[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58]
[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78]


## 参数

In [165]:
def greet(name='', msg=''):
    print("Hello", name + ', ' + msg)


greet("Monica", "Good morning!")

Hello Monica, Good morning!


In [162]:
name='Monica'
print("Hello", name + ', ' + msg)

NameError: name 'msg' is not defined

In [168]:
greet(msg='Good morning!',name='Monica')

Hello Monica, Good morning!


In [38]:
greet("Monica")

greet()

TypeError: greet() missing 1 required positional argument: 'msg'

## 变量函数参数

### Python默认参数

In [39]:
def greet(name, msg="Good morning!"):
    print("Hello", name + ', ' + msg)


greet("Kate")
greet("Bruce", "How do you do?")

Hello Kate, Good morning!
Hello Bruce, How do you do?


### Python关键字参数

In [41]:
greet(name="Bruce", msg="How do you do?")
greet(msg="How do you do?", name="Bruce")
greet("Bruce", msg="How do you do?")

Hello Bruce, How do you do?
Hello Bruce, How do you do?
Hello Bruce, How do you do?


### Python任意参数

In [42]:
def greet(*names):
    for name in names:
        print("Hello", name)


greet("Monica", "Luke", "Steve", "John")

Hello Monica
Hello Luke
Hello Steve
Hello John


## Python函数: 在列表中移除重复对象

In [43]:
def remove_duplicates(duplicate):
    uniques = []
    for num in duplicate:
        if num not in uniques:
            uniques.append(num)
    return(uniques)


duplicate = [2, 4, 10, 20, 5, 2, 20, 4]
print(remove_duplicates(duplicate))

[2, 4, 10, 20, 5]


# 面向对象编程

## 类
物以类聚

## 对象（实例）
消费的会员的数据，会员九可以当作我们数据处理的对象，这个对象（会员）一般有一些基本的信息，比如年龄、消费金额、消费时间等等。

对象的信息放到程序里（项目），就需要将对象抽象成类。对象抽象成了一个具有一些基本属性的可以描述的语言，一些技能就需要用方法来描述

## 定义类

In [1]:
class Dog:
    pass

## 实例属性

In [12]:
class Dog:
    def __init__(self, name, age): # 构造方法，我要养狗的话，狗必须要有个名字，狗肯定有个年龄的属性
        self.name = name
        self.age = age

In [11]:
dog = Dog(name='小花', age=1)

TypeError: __init__() missing 1 required positional argument: 'age'

## 类属性

In [22]:
class Dog:
    # Class Attribute
    species = 'mammal' # 物种是哺乳动物

    # Initializer / Instance Attributes
    def __init__(self, name, age):
        self.name = name
        self.age = age

In [9]:
dog = Dog(name='小花', age=1)

In [14]:
dog.species='哺乳动物'

In [15]:
dog.species

'哺乳动物'

## 实例化对象

In [16]:
class Dog:
    pass

In [17]:
Dog()

In [18]:
Dog()

In [19]:
a = Dog()
b = Dog()
a == b

False

In [21]:
type(b)

__main__.Dog

In [23]:
a = Dog(name='小花', age=1)
b = Dog(name='小白', age=3)

In [25]:
a.name,b.name,b.species,a.species

('小花', '小白', 'mammal', 'mammal')

In [31]:
class Dog:
    # Class Attribute
    species = 'mammal'

    # Initializer / Instance Attributes
    def __init__(self, name, age):
        self.name = name
        self.age = age


# Instantiate the Dog object
philo = Dog("Philo", 5)
mikey = Dog("Mikey", 6)

# Access the instance attributes
print("{} is {} and {} is {}.".format(philo.name, philo.age, mikey.name, mikey.age))

# Is Philo a mammal?
if philo.species == "mammal":
    print("{0} is a {1}!".format(philo.name, philo.species))

Philo is 5 and Mikey is 6.
Philo is a mammal!


## 实例方法

In [32]:
class Dog:

    # Class Attribute
    species = 'mammal' # 通用属性

    # Initializer / Instance Attributes
    def __init__(self, name, age):
        self.name = name # 私有属性
        self.age = age

    # instance method
    def description(self):
        return "{} is {} years old".format(self.name, self.age)

    # instance method
    def speak(self, sound):
        return "{} says {}".format(self.name, sound)


# Instantiate the Dog object
mikey = Dog("Mikey", 6)

# call our instance methods
print(mikey.description())
print(mikey.speak("Gruff Gruff"))

Mikey is 6 years old
Mikey says Gruff Gruff


In [33]:
hua=Dog('小花',4)

In [51]:
print(hua.speak('汪汪'))

小花 says 汪汪


In [52]:
mikey.speak('汪汪')

'Mikey says 汪汪'

In [54]:
hua.name='小白'

In [55]:
hua.name

'小白'

### 修改属性

In [74]:
class Email:
    is_get = '收到了邮件'

    def __init__(self):
        self.is_sent = False

    def send_email(self):
        self.is_sent = True
        
    def get_email(self):
        is_get = '收到'
        return is_get


my_email = Email()
print(my_email.is_sent)

my_email.send_email()
print(my_email.is_sent)

False
True


In [62]:
print(my_email.is_sent)
my_email.is_sent = False
print(my_email.is_sent)
my_email.send_email()
print(my_email.is_sent)

True
False
True


In [75]:
print(my_email.is_get)
print(my_email.get_email())
print(my_email.is_get)

收到了邮件
收到
收到了邮件


In [81]:
class Da():

    def add(self, a, b):
        return a+b

    def mins(self, a, b):
        return a-b


da = Da()
# print(add(10, 20))
# print(mins(10, 20))

In [82]:
print(da.add(10,20))

30


## Python对象继承

In [83]:
class Dog(object):
    pass


class Dog:
    pass

## 狗公园示例

In [84]:
class Dog:
    def __init__(self, breed):
        self.breed = breed


spencer = Dog("German Shepard")
print(spencer.breed)

sara = Dog("Boston Terrier")
print(sara.breed)

German Shepard
Boston Terrier


In [85]:
class Spencer(Dog):
    def speek(self):
        return '我来自德国'

In [89]:
ss = Spencer('German')

In [88]:
spencer.breed

'German Shepard'

In [90]:
ss.breed

'German'

In [91]:
ss.speek()

'我来自德国'

### 扩展父类的功能

In [92]:
class Dog:

    # Class attribute
    species = 'mammal'

    # Initializer / Instance attributes
    def __init__(self, name, age):
        self.name = name
        self.age = age

    # instance method
    def description(self):
        return "{} is {} years old".format(self.name, self.age)

    # instance method
    def speak(self, sound):
        return "{} says {}".format(self.name, sound)


# Child class (inherits from Dog class)
class RussellTerrier(Dog):
    def run(self, speed):
        return "{} runs {}".format(self.name, speed)


# Child class (inherits from Dog class)
class Bulldog(Dog):
    def run(self, speed):
        return "{} runs {}".format(self.name, speed)


# Child classes inherit attributes and
# behaviors from the parent class
jim = Bulldog("Jim", 12)
print(jim.description())

# Child classes have specific attributes
# and behaviors as well
print(jim.run("slowly"))

Jim is 12 years old
Jim runs slowly


In [96]:
class 马:
    pass


class 驴:
    pass


class 骡子(马, 驴):
    pass

In [93]:
jim.run('快')

'Jim runs 快'

## 父类与子类
isinstance（）函数用于确定实例是否也是某个父类的实例。

In [61]:
# Parent class
class Dog:

    # Class attribute
    species = 'mammal'

    # Initializer / Instance attributes
    def __init__(self, name, age):
        self.name = name
        self.age = age

    # instance method
    def description(self):
        return "{} is {} years old".format(self.name, self.age)

    # instance method
    def speak(self, sound):
        return "{} says {}".format(self.name, sound)


# Child class (inherits from Dog() class)
class RussellTerrier(Dog):
    def run(self, speed):
        return "{} runs {}".format(self.name, speed)


# Child class (inherits from Dog() class)
class Bulldog(Dog):
    def run(self, speed):
        return "{} runs {}".format(self.name, speed)


# Child classes inherit attributes and
# behaviors from the parent class
jim = Bulldog("Jim", 12)
print(jim.description())

# Child classes have specific attributes
# and behaviors as well
print(jim.run("slowly"))

# Is jim an instance of Dog()?
print(isinstance(jim, Dog))

# Is julie an instance of Dog()?
julie = Dog("Julie", 100)
print(isinstance(julie, Dog))

# Is johnny walker an instance of Bulldog()
johnnywalker = RussellTerrier("Johnny Walker", 4)
print(isinstance(johnnywalker, Bulldog))

# Is julie and instance of jim?
print(isinstance(julie, jim))

Jim is 12 years old
Jim runs slowly
True
True
False


TypeError: isinstance() arg 2 must be a type or tuple of types

### 覆盖父类的功能

In [97]:
class Dog:
    species = 'mammal'


class SomeBreed(Dog):
    pass


class SomeOtherBreed(Dog):
    species = 'reptile'


frank = SomeBreed()
print(frank.species)

beans = SomeOtherBreed()
print(beans.species)

mammal
reptile


# 练习

## 将字符串转化成小写字母

In [63]:
def toLowerCase(str: str):
    str = str.lower()
    return str


toLowerCase('Hello World!')

'hello world!'

## 反转字符串中的单词

In [88]:
def reverseWords(s: str):
    word = []
    s = s.split(' ')
    for i in s:
        word.append(i[::-1])
    return ' '.join(word)


reverseWords('Let\'s take deep learning contest')

"s'teL ekat peed gninrael tsetnoc"

In [17]:
def reverseWords(s: str):
    word = [] # 定义空的列表，目的是为了保存逆序的字符
    s = s.split(' ') # 将输入的字符串用空格分割成单词列表
#     print(s)
    for i in s: # 遍历单词列表
#         print(i)
        word.append(i[::-1]) # 字符串反向切片
#     print(word)
    return ' '.join(word)


reverseWords('Let\'s take deep learning contest')

["s'teL", 'ekat', 'peed', 'gninrael', 'tsetnoc']


"s'teL ekat peed gninrael tsetnoc"

In [18]:
'Let\'s take deep learning contest'[::-1]

"tsetnoc gninrael peed ekat s'teL"

## 排列组合
有四个数字：1、2、3、4，能组成多少个互不相同且无重复数字的三位数？各是多少？

In [95]:
nums = []
for i in range(1, 5):
    for j in range(1, 5):
        if j == i:
            continue
        for k in range(1, 5):
            if k == i or k == j:
                continue
            num = i*100+j*10+k
            if num not in nums:
                nums.append(num)
print(len(nums), nums)

24 [123, 124, 132, 134, 142, 143, 213, 214, 231, 234, 241, 243, 312, 314, 321, 324, 341, 342, 412, 413, 421, 423, 431, 432]


## 阶乘
让函数FirstFactorial（num）接受传递的num参数并返回它的阶乘。例如：如果num = 4，则程序应返回（4 * 3 * 2 * 1）= 24.对于测试用例，范围将介于1和18之间，输入将始终为整数

In [99]:
def FirstFactorial(num):
    n = 1
    for i in range(1, num+1):
        n = n*i
    return n


print(FirstFactorial(18))

6402373705728000


## 整型转罗马数字
罗马数字 3 写做 III ，即为3个并列的 1。12 写做 XII ，即为 X + II 。 26 写做 XXVI, 即为 XX + V + I 。 也有一些特例, 例如4, 可以放到V的左边, IV表示4.

In [112]:
def intToRoman(num: int):
    x = int(num/10) # 除以10后取整数部分，确定数字有多少个X
    y = num % 10 # 取数字的个位数
    if y < 5:
        if y < 4:
            roman = 'X'*x+'I'*y
        else:
            roman = 'X'*x+'I'+'V'
    elif y == 5:
        roman = 'X'*x+'V'
    elif y == 9:
        roman = 'X'*x+'I'+'X'
    else:
        roman = 'X'*x+'V'+'I'*(y % 5)
    return roman


print(intToRoman(115))

XXXXXXXXXXXV


In [25]:
def intToRoman(num: int):
    x = int(num/10) # 除以10后取整数部分，确定数字有多少个X
    y = num % 10 # 取数字的个位数
    if y == 1:
        z = 'I'
    elif y == 2:
        z = 'II'
    elif y == 3:
        z = 'III'
    elif y == 4:
        z = 'IV'
    elif y == 5:
        z = 'V'
    elif y == 6:
        z = 'VI'
    elif y == 7:
        z = 'VII'
    elif y == 8:
        z = 'VIII'
    elif y == 9:
        z = 'IX'
    elif y == 0:
        z = ''
    return 'X'*x+z


print(intToRoman(25))

XXV


In [21]:
'X'*int(25/10)

'XX'

解决问题需要分哪几种情况
* 第一个，十进制，那么大于10的部分就是，54就代表有5个X加上一个IV，第一步确定有多少个X；
* 第二个，I,II,III,IV,V,VI,VII,VIII,IX,X

## 编程实现

一、猜数字
经典的猜数字游戏，几乎所有人学编程时都会做。

功能描述：随机选择一个三位以内的数字作为答案。用户输入一个数字，程序会提示大了或是小了，直到用户猜中。

二、FizzBuzz
另一道经典编程题。

功能描述：遍历并打印0到100，如果数字能被3整除，显示Fizz；如果数字能被5整除，显示Buzz；如果能同时被3和5整除，就显示FizzBuzz。结果应该类似：0,1,2，Fizz，4，Buzz，6……14，FizzBuzz，16……

三、猜数字的AI
和猜数字一样，不过这次是设计一个能猜数字的AI

功能描述：用户输入一个单位以内的数字，AI要用最少的次数猜中，并且显示出猜的次数和数字。

四、整点报时
老式挂钟会在整点的报时，响铃的次数和时间相等。我们设计一个在电脑上运行的报时器。

功能描述：运行后，在每一个整点长响一声，半个整点短响两声。实现睡眠模式，晚上十二点到早上六点不响铃。

五、敏感词审查
一向痛恨网络审查的你成了某网站的审查员，要求审查网络上违反上头规定的名词。

功能描述：要审查的帖子在这个文本文档里，要求将所有的和谐，三个代表，言论自由，64替换为*号。

六、文字冒险游戏
你想做个清新脱俗的文字游戏，只在命令行的黑框里运行。

功能描述：尽量参考回合制RPG游戏的模式，游戏剧情自编，要有战斗模式，主角有HP属性，战斗和吃东西会对HP有相应的增减效果。

七、词频统计
你想分析某个英语写作者的写作风格，你想先从他使用词语的频率入手。

功能描述：分析这个文本文档里的词频，按照出现频率由高到低排列结果，不区分大小写，过滤掉标点（可以使用正则表达式）。结果类似ok：234，play：122，funny：78

八、Markdown转换器
你一向喜欢在Windows的记事本里写文章，为了能方便的生成网页，你使用Markdown的格式写了篇文章，现在需要一个把Markdown转换成HTML格式的程序。

功能描述：将这个文本文档里的内容转换成HTML文档。要尽可能的支持Markdown标签，越多越好。建议使用正则表达式。

九、邮件轰炸机
你的一个朋友生日快到了，你决定搞个恶作剧——用邮件塞满他的邮箱。

功能描述：首先去注册十个邮箱，然后用这是个邮箱轮流发送随机生成的邮件内容给你的朋友。调整发送的频率，以免被服务器拒绝。

十、用邮件控制电脑
你的电脑放在宿舍，这时你在教室上课，你又决定搞个恶作剧吓吓装病请假在宿舍玩游戏的室友。

功能描述：使用邮件发送特定的命令，让电脑上的程序判断发信方是否为控制邮箱，如果是就执行发送过来的命令。通过不同的方式实现尽可能多的功能。